adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -735,41 +827,143 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+   *
+   * @note This is best used under the condition that (repeatedly)
+   * repartitioning is more expensive than calculating the proper location
+   * of the records of this.rdd. It requires a pass through the co-located
+   * RDD to get the correct partition(s) for each record. It will assign a
+   * record to multiple partitions if necessary.
+   * @param rddToCoPartitionWith The rdd to copartition to.
+   * @return The newly repartitioned rdd.
+   */
+  private[rdd] def copartitionByReferenceRegion[X, Y <: GenomicRDD[X, Y]](
+    rddToCoPartitionWith: GenomicRDD[X, Y])(implicit tTag: ClassTag[T], xTag: ClassTag[X]): U = {
+
+    // if the other RDD is not sorted, we can't guarantee proper copartition
+    assert(rddToCoPartitionWith.sorted,
+      "Cannot copartition with an unsorted rdd!")
+
+    val destinationPartitionMap = rddToCoPartitionWith.optPartitionMap.get
+    //number of partitions we will have after repartition
+    val numPartitions = destinationPartitionMap.length
+    // here we create a partition map with a single ReferenceRegion that spans
+    // the entire range, however we have to handle the case where the partition
+    // spans multiple referenceNames because of load balancing.
+    val adjustedPartitionMapWithIndex =
+      // the zipWithIndex gives us the destination partition ID
+      destinationPartitionMap.flatten.zipWithIndex.map(g => {
+        val (firstRegion, secondRegion, index) = (g._1._1, g._1._2, g._2)
+        // in the case where we span multiple referenceNames
+        if (firstRegion.referenceName != secondRegion.referenceName) {
+          // the first region is enough to represent the partition
+          (firstRegion, index)
+        } else {
+          // otherwise we just have the ReferenceRegion span from partition
+          // start to end
+          (ReferenceRegion(
+            firstRegion.referenceName,
+            firstRegion.start,
+            secondRegion.end),
+            index)
+        }
       })
-      .asInstanceOf[GenomicRDD[(Option[T], Iterable[X]), Z]]
+
+    // convert to an IntervalArray for fast range query
+    val partitionMapIntervals = IntervalArray(
+      adjustedPartitionMapWithIndex,
+      adjustedPartitionMapWithIndex.maxBy(_._1.width)._1.width,
+      sorted = true)
+
+    val finalPartitionedRDD = {
+      // if not already sorted, we need to duplicate those records that are
+      // mapped to multiple ReferenceRegions, thus we use the flattenRddByRegions
+      val referenceRegionKeyedGenomicRDD = flattenRddByRegions()
+
+      referenceRegionKeyedGenomicRDD.mapPartitions(iter => {
+        iter.flatMap(f => {
+          val intervals = partitionMapIntervals.get(f._1, requireOverlap = false)
+          intervals.map(g => ((f._1, g._2), f._2))
+        })
+      }, preservesPartitioning = true)
+        .repartitionAndSortWithinPartitions(
+          ManualRegionPartitioner(numPartitions))
+    }
+    finalPartitionedRDD.cache()
```

Comment:
We likely shouldn't be. Previously we were recomputing a new `optPartitionMap` but we didn't need to, this is an artifact of that.